sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala

```diff
@@ -91,6 +90,25 @@ case class ResolveDefaultColumns(catalog: SessionCatalog) extends Rule[LogicalPl
     }
   }
 
+  /**
+   * Checks if a logical plan is an INSERT INTO command where the inserted data comes from a SELECT
+   * list, with possible other unary operators like sorting and/or alias(es) in between.
+   */
+  private def insertsFromProject(i: InsertIntoStatement): Option[Project] = {
+    var node = i.query
+    def matches(node: LogicalPlan): Boolean = node match {
+      case _: GlobalLimit | _: LocalLimit | _: Offset | _: SubqueryAlias | _: Sort => true
+      case _ => false
+    }
+    while (matches(node)) {
+      node = node.children.head
```

Comment:
Why getting the first child only? What if there is a join and one of the children has a project containing default?

Reply:
Good question, I added a test for this:

```
insert into t2 (i, s) select default, default from t1 inner join t1 using (i, s);

> true, 42L, "abc",
> true, 42L, "abc"
```

We want the `default` resolution to cover the topmost `select` list before the join, but not any below, per the specification. For example, this should not work:

```
insert into t2 (i, s) select i, s from (
  select default as i, default as s from t1)
  inner join t1
  using (i, s);
```